std/digest/murmurhash.d

```diff
@@ -0,0 +1,860 @@
+alias MurmurHash3_128_opt32Digest = WrapperDigest!MurmurHash3_128_opt32;
+/// MurmurHash3_128_opt64 $(D std.digest.digest.Digest) OOO API.
+alias MurmurHash3_128_opt64Digest = WrapperDigest!MurmurHash3_128_opt64;
+
+/*
+Performance notes:
+ - To help a bit with the performance when compiling with DMD some functions in
+   this module have been duplicated, some other functions have been rewritten to
+   pass by value instead of by reference.
+ - GDC and LDC are on par with their C++ counterpart.
+ - DMD is typically between 20% to 50% of the GCC version.
+ -
+*/
+
+/++
+MurmurHash3 optimized for x86 processors producing a 32-bit value.
+
+This is a lower level implementation that makes finalization optional and have
+better performance than $(D digest).
+Note that $(D putRemainder) can be called only once and that no subsequent calls
+to $(D putBlocks) is allowed.
++/
+struct FastMurmurHash3_32_opt32
+{
+private:
+    enum uint c1 = 0xcc9e2d51;
+    enum uint c2 = 0x1b873593;
+    uint h1;
+
+public:
+    alias Block = uint; /// The element type for 32-bit implementation.
+    size_t size;
+
+    this(uint seed)
+    {
+        h1 = seed;
+    }
+
+    @disable this(this);
+
+    /++
+    Adds a single Block of data without increasing size.
+    Make sure to increase size by Block.sizeof for each call to putBlock.
+    +/
+    void putBlock(uint block) pure nothrow @nogc
+    {
+        h1 = update(h1, block, 0, c1, c2, 15, 13, 0xe6546b64U);
+    }
+
+    /// Put remainder bytes. This must be called only once after putBlock and before finalize.
+    void putRemainder(scope const(ubyte[]) data...) pure nothrow @nogc
+    {
+        assert(data.length < Block.sizeof);
+        assert(data.length >= 0);
+        size += data.length;
+        uint k1 = 0;
+        final switch (data.length & 3)
+        {
+        case 3:
+            k1 ^= data[2] << 16;
+            goto case;
+        case 2:
+            k1 ^= data[1] << 8;
+            goto case;
+        case 1:
+            k1 ^= data[0];
+            h1 ^= shuffle(k1, c1, c2, 15);
+            goto case;
+        case 0:
+        }
+    }
+
+    /// Incorporate size and finalizes the hash.
+    void finalize() pure nothrow @nogc
+    {
+        h1 ^= size;
+        h1 = fmix(h1);
+    }
+
+    /// Returns the hash as an uint value.
+    Block get() pure nothrow @nogc
+    {
+        return h1;
+    }
+
+    /++
+    Pushes an array of blocks at once. It is more efficient to push as much data as possible in a single call.
+    On platform that does not support unaligned reads (MIPS or old ARM chips), the compiler may produce slower code to ensure correctness.
+    +/
+    void putBlocks(scope const(Block[]) blocks...) pure nothrow @nogc
+    {
+        foreach (const block; blocks)
+        {
+            putBlock(block);
+        }
+        size += blocks.length * Block.sizeof;
+    }
+
+    /// Returns the current hashed value as an ubyte array.
+    auto getBytes() pure nothrow @nogc
```

Comment:
Would be probably nice to specify the return type so it appears in clear to the reader.